ocaml/xapi/certificates.ml

```diff
@@ -334,3 +335,57 @@ let validate_private_key pkcs8_private_key =
   |> function
     | Ok priv -> priv
     | Error `Msg (err, msg) -> raise_server_error msg err
+
+let validate_certificate kind str now private_key =
+  let open Rresult in
+  let ensure_keys_match private_key certificate =
+    let public_key = X509.Certificate.public_key certificate in
+    match public_key, private_key with
+    | `RSA pub, `RSA priv when pub = Nocrypto.Rsa.pub_of_priv priv ->
+        Ok certificate
+    | _ ->
+        Error (`Msg (server_certificate_key_mismatch, []))
+  in
+
+  let ensure_validity ~time certificate =
+    let to_string = Ptime.to_rfc3339 in
```

Comment:
Should we also include some grace period here?
i.e. if someone tries to install a certificate that expires the next day we should probably warn them in that case too?
i.e. have a minimum validity period of lets say 30 days?

Reply:
There are alerts planned to warn users about expiring server certificates, I'm not sure if it's worth doing the check here as well. I'll take note and see if it's easy to reuse those checks here as well.